File: app/src/main/java/com/example/clubmanager/ui/ping/ComposePingActivity.java
package com.example.clubmanager.ui.ping;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.comp3074uiwireframe.R;
import com.example.clubmanager.data.models.Ping;
import com.example.clubmanager.data.models.TargetType;
import com.example.clubmanager.data.models.Urgency;
import com.example.clubmanager.data.repo.ServiceLocator;

import java.util.ArrayList;
import java.util.UUID;

public class ComposePingActivity extends AppCompatActivity {

    // 🔹 UI fields
    private RadioGroup rgTarget;
    private RadioButton rbAll, rbRole, rbStaff;
    private EditText etRole, etStaffIds, etMessage;
    private Spinner spUrgency;
    private Button btnSend;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_compose_ping);

        // ✅ Enable back arrow + title
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle("Compose Ping");
        }

        // ✅ Hook up views
        rgTarget = findViewById(R.id.rgTarget);
        rbAll = findViewById(R.id.rbAll);
        rbRole = findViewById(R.id.rbRole);
        rbStaff = findViewById(R.id.rbStaff);

        etRole = findViewById(R.id.etRole);
        etStaffIds = findViewById(R.id.etStaffIds);
        etMessage = findViewById(R.id.etMessage);

        spUrgency = findViewById(R.id.spUrgency);
        btnSend = findViewById(R.id.btnSend);

        // ✅ Populate urgency spinner (Normal / High / Low)
        ArrayAdapter<String> urgencyAdapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                new String[]{"Normal", "High", "Low"}
        );
        urgencyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spUrgency.setAdapter(urgencyAdapter);
        spUrgency.setSelection(0); // default: Normal

        // ✅ Show/hide Role / Staff ID fields based on radio selection
        rgTarget.setOnCheckedChangeListener((group, checkedId) -> {
            if (checkedId == R.id.rbRole) {
                etRole.setVisibility(View.VISIBLE);
                etStaffIds.setVisibility(View.GONE);
            } else if (checkedId == R.id.rbStaff) {
                etRole.setVisibility(View.GONE);
                etStaffIds.setVisibility(View.VISIBLE);
            } else {
                // "All staff"
                etRole.setVisibility(View.GONE);
                etStaffIds.setVisibility(View.GONE);
            }
        });

        // Ensure default state (All staff is checked in XML)
        etRole.setVisibility(View.GONE);
        etStaffIds.setVisibility(View.GONE);

        // ✅ Send button logic using real form values
        btnSend.setOnClickListener(v -> {
            // 1) Validate message
            String message = etMessage.getText().toString().trim();
            if (message.isEmpty()) {
                etMessage.setError("Please enter a message");
                etMessage.requestFocus();
                return;
            }

            // 2) Determine target type + recipients
            int checkedId = rgTarget.getCheckedRadioButtonId();
            TargetType targetType;
            ArrayList<String> toIds = new ArrayList<>();

            if (checkedId == R.id.rbRole) {
                // Send by role
                String role = etRole.getText().toString().trim();
                if (role.isEmpty()) {
                    etRole.setError("Please enter a role");
                    etRole.requestFocus();
                    return;
                }
                targetType = TargetType.ROLE;
                // For now we just store the role name in toIds
                toIds.add(role);

            } else if (checkedId == R.id.rbStaff) {
                // Send to specific staff IDs
                String idsText = etStaffIds.getText().toString().trim();
                if (idsText.isEmpty()) {
                    etStaffIds.setError("Please enter at least one staff ID");
                    etStaffIds.requestFocus();
                    return;
                }

                targetType = TargetType.STAFF;
                // Split by comma: "s1, s2, s3"
                String[] parts = idsText.split(",");
                for (String p : parts) {
                    String id = p.trim();
                    if (!id.isEmpty()) {
                        toIds.add(id);
                    }
                }
                if (toIds.isEmpty()) {
                    etStaffIds.setError("Please enter valid staff IDs");
                    etStaffIds.requestFocus();
                    return;
                }

            } else {
                // All staff
                targetType = TargetType.ALL;
                // toIds stays empty
            }

            // 3) Read urgency from spinner
            String selectedUrgency = (String) spUrgency.getSelectedItem();
            Urgency urgency;
            if ("High".equalsIgnoreCase(selectedUrgency)) {
                urgency = Urgency.HIGH;
            } else if ("Low".equalsIgnoreCase(selectedUrgency)) {
                urgency = Urgency.LOW;
            } else {
                urgency = Urgency.NORMAL;
            }

            // 4) Build Ping object and save via repo
            String id = UUID.randomUUID().toString();
            long now = System.currentTimeMillis();

            Ping ping = new Ping(
                    id,
                    targetType,
                    toIds,
                    message,
                    urgency,
                    now,
                    new ArrayList<>()   // nobody has acknowledged yet
            );

            try {
                ServiceLocator.repo.addPing(ping);
                Toast.makeText(this, "Ping sent!", Toast.LENGTH_SHORT).show();
                finish(); // Close after sending
            } catch (Exception e) {
                Toast.makeText(this, "Error: " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        });
    }

    // ✅ Handle back arrow click
    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/com/example/clubmanager/ui/ping/ComposePingActivity.kt
package com.example.clubmanager.ui.ping

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.example.clubmanager.R
import com.example.clubmanager.data.models.Ping
import com.example.clubmanager.data.models.TargetType
import com.example.clubmanager.data.models.Urgency
import com.example.clubmanager.data.repo.ServiceLocator
import java.util.UUID
import android.content.DialogInterface


class ComposePingActivity : AppCompatActivity() {

    private lateinit var rgTarget: RadioGroup
    private lateinit var rgUrgency: RadioGroup
    private lateinit var rbAll: RadioButton
    private lateinit var rbRole: RadioButton
    private lateinit var rbStaff: RadioButton
    private lateinit var rbLow: RadioButton
    private lateinit var rbNormal: RadioButton
    private lateinit var rbHigh: RadioButton
    private lateinit var etRole: EditText
    private lateinit var etMessage: EditText
    private lateinit var tvStaffLabel: TextView
    private lateinit var btnPickStaff: Button
    private lateinit var btnSend: Button
    private lateinit var btnOpenInbox: Button

    private val selectedStaffIds = mutableListOf<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_compose_ping)

        rgTarget = findViewById(R.id.rgTarget)
        rgUrgency = findViewById(R.id.rgUrgency)
        rbAll = findViewById(R.id.rbAll)
        rbRole = findViewById(R.id.rbRole)
        rbStaff = findViewById(R.id.rbStaff)
        rbLow = findViewById(R.id.rbLow)
        rbNormal = findViewById(R.id.rbNormal)
        rbHigh = findViewById(R.id.rbHigh)
        etRole = findViewById(R.id.etRole)
        etMessage = findViewById(R.id.etMessage)
        tvStaffLabel = findViewById(R.id.tvStaffLabel)
        btnPickStaff = findViewById(R.id.btnPickStaff)
        btnSend = findViewById(R.id.btnSend)
        btnOpenInbox = findViewById(R.id.btnOpenInbox)

        rgTarget.setOnCheckedChangeListener { _, _ -> updateVisibility() }
        updateVisibility()

        btnPickStaff.setOnClickListener { showStaffPicker() }
        btnSend.setOnClickListener { onSend() }
        btnOpenInbox.setOnClickListener {
            startActivity(Intent(this, InboxActivity::class.java))
        }
    }

    private fun updateVisibility() {
        val byRole = rbRole.isChecked
        val byStaff = rbStaff.isChecked
        etRole.visibility = if (byRole) View.VISIBLE else View.GONE
        tvStaffLabel.visibility = if (byStaff) View.VISIBLE else View.GONE
        btnPickStaff.visibility = if (byStaff) View.VISIBLE else View.GONE
    }

    private fun showStaffPicker() {
        val staff = ServiceLocator.repo.staff
        val names: Array<CharSequence> =
            staff.map { "${it.name} (${it.role})" }.toTypedArray()
        val checked: BooleanArray =
            staff.map { selectedStaffIds.contains(it.id) }.toBooleanArray()

        AlertDialog.Builder(this)
            .setTitle("Select staff")
            .setMultiChoiceItems(names, checked) { _: DialogInterface, which: Int, isChecked: Boolean ->
                val id = ServiceLocator.repo.staff[which].id
                if (isChecked) {
                    if (!selectedStaffIds.contains(id)) selectedStaffIds.add(id)
                } else {
                    selectedStaffIds.remove(id)
                }
            }
            .setPositiveButton("Done") { dialog: DialogInterface, _: Int -> dialog.dismiss() }
            .show()


    }

    private fun onSend() {
        val target = when {
            rbRole.isChecked -> TargetType.ROLE
            rbStaff.isChecked -> TargetType.STAFF
            else -> TargetType.ALL
        }

        val urgency = when {
            rbHigh.isChecked -> Urgency.HIGH
            rbLow.isChecked -> Urgency.LOW
            else -> Urgency.NORMAL
        }

        val msg = etMessage.text?.toString()?.trim().orEmpty()
        if (TextUtils.isEmpty(msg)) {
            Toast.makeText(this, "Message required", Toast.LENGTH_SHORT).show()
            return
        }

        val toIds = when (target) {
            TargetType.ALL -> emptyList()
            TargetType.ROLE -> {
                val role = etRole.text?.toString()?.trim().orEmpty()
                if (role.isEmpty()) {
                    Toast.makeText(this, "Enter a role", Toast.LENGTH_SHORT).show()
                    return
                }
                listOf(role)
            }
            TargetType.STAFF -> {
                if (selectedStaffIds.isEmpty()) {
                    Toast.makeText(this, "Pick at least one staff", Toast.LENGTH_SHORT).show()
                    return
                }
                selectedStaffIds.toList()
            }
        }

        val ping = Ping(
            id = UUID.randomUUID().toString(),
            toType = target,
            toIds = toIds,
            message = msg,
            urgency = urgency
        )
        ServiceLocator.repo.addPing(ping)
        Toast.makeText(this, "Ping sent!", Toast.LENGTH_SHORT).show()

        etMessage.setText("")
        selectedStaffIds.clear()
    }
}
